src/components/Widget/components/Conversation/components/Sender/index.js

```diff
@@ -2,20 +2,167 @@ import React from 'react';
 import PropTypes from 'prop-types';
 
 import send from 'assets/send_button.svg';
+import micOn from 'assets/microphone.svg';
+import micOff from 'assets/microphone-slash.svg';
 import './style.scss';
 
-const Sender = ({ sendMessage, placeholder, disabledInput }) =>
-  <form className="sender" onSubmit={sendMessage}>
-    <input type="text" className="new-message" name="message" placeholder={placeholder} disabled={disabledInput} autoFocus autoComplete="off" />
-    <button type="submit" className="send">
-      <img src={send} className="send-icon" alt="send" />
-    </button>
-  </form>;
+export default class Sender extends React.Component {
+  constructor() {
+    super();
+    this.state = {
+      recognizing: false,
+      icon: micOn,
+      color: 'blue',
+      interim_transcript: '',
+      final_transcript: '',
+      ignore_onend: null,
+      start_timestamp: null
+    };
+  }
+  componentDidMount() {
+    const two_line = /\n\n/g;
+    const one_line = /\n/g;
+    function linebreak(s) {
+      return s.replace(two_line, '<p></p>').replace(one_line, '<br>');
+    }
+
+    const first_char = /\S/;
+    function capitalize(s) {
+      return s.replace(first_char, m => m.toUpperCase());
+    }
+
+    function upgrade() {
+      console.log('Web Speech API is not supported by this browser. Upgrade to Chrome version 25 or later.');
+    }
+
+
+    if (!('webkitSpeechRecognition' in window)) {
+      upgrade();
+    } else {
+      this.recognition = new webkitSpeechRecognition();
+      this.recognition.continuous = true;
+      this.recognition.interimResults = true;
+
+      this.recognition.onstart = () => {
+        this.setState({
+          recognizing: true
+        });
+        console.log('Speak now.');
+      };
+
+      this.recognition.onerror = (event) => {
+        console.error(event.error);
+        if (event.error === 'no-speech') {
+          console.error('No speech was detected. You may need to adjust your microphone settings');
+          this.setState({
+            ignore_onend: true
```

Comment:
This `setState` is being repeated in the 3 cases. I'd set the `ignore_onend` outside the ifs and just asking
```js
if (event.error) this.setState({ ignore_onend: true });
```